### ISSUE TITLE
Nmemory update error with more 10000 items

### ISSUE THREAD

**Issue by Bykiev** [zzzprojects/nmemory#44]
@JonathanMagnan, hi!

I'm using Effort project for unit testing, I have an array of test data (about 25k rows), but NMemory throw update error, if I'll add by chanks (call SaveChanges after each 5000 records) all is working fine.

StackTrace:
```
NMemory.Transactions.Transaction.EnsureTransaction(Transaction& transaction, IDatabase database)
   в NMemory.Tables.Table`2.Insert(TEntity entity, Transaction transaction)
   в Effort.Internal.Common.DatabaseReflectionHelper.WrapperMethods.InsertEntity[TEntity](ITable`1 table, TEntity entity, Transaction transaction)
--- Конец трассировки стека внутренних исключений ---
    в System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   в System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   в System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   в Effort.Internal.Common.DatabaseReflectionHelper.InsertEntity(ITable table, Object entity, Transaction transaction)
   в Effort.Internal.CommandActions.InsertCommandAction.CreateAndInsertEntity(ITable table, IList`1 memberBindings, Transaction transaction)
   в Effort.Internal.CommandActions.InsertCommandAction.ExecuteNonQuery(ActionContext context)
   в Effort.Provider.EffortEntityCommand.ExecuteNonQuery()
   в System.Data.Entity.Infrastructure.Interception.DbCommandDispatcher.<>c.<NonQuery>b__4_0(DbCommand t, DbCommandInterceptionContext`1 c)
   в System.Data.Entity.Infrastructure.Interception.InternalDispatcher`1.Dispatch[TTarget,TInterceptionContext,TResult](TTarget target, Func`3 operation, TInterceptionContext interceptionContext, Action`3 executing, Action`3 executed)
   в System.Data.Entity.Infrastructure.Interception.DbCommandDispatcher.NonQuery(DbCommand command, DbCommandInterceptionContext interceptionContext)
   в System.Data.Entity.Internal.InterceptableDbCommand.ExecuteNonQuery()
   в System.Data.Entity.Core.Mapping.Update.Internal.DynamicUpdateCommand.Execute(Dictionary`2 identifierValues, List`1 generatedValues)
   в System.Data.Entity.Core.Mapping.Update.Internal.UpdateTranslator.Update()
```

**Comment by Bykiev** [comment]
I found the same issue and after setting ConnectionTimeout all is working as expected. You should review the code and improve exception handling.

**Comment by JonathanMagnan** [comment]
Hello @Bykiev ,
 
Do you think you could provide a **runnable project sample** for this issue? It will help my developer to get started to investigate it faster and make sure nothing is missing.

_Providing a project sample is now **REQUIRED**. It happened **too many times** that something **was missing** to investigate and/or answer an issue._

_Try to create a new project with only the minimal code (having too many non-related codes doesn’t help either)._  

You can send it to info@zzzprojects.com if you need to keep the source private

Best Regards,
 
Jon
 
---
 
**Performance Libraries**
`context.BulkInsert(list, options => options.BatchSize = 1000);  `
[Entity Framework Extensions](http://entityframework-extensions.net/) • [Entity Framework Classic](http://entityframework-classic.net/) • [Bulk Operations](http://bulk-operations.net/) •  [Dapper Plus](http://dapper-plus.net/)
 
**Runtime Evaluation**
`Eval.Execute("x + y", new {x = 1, y = 2}); // return 3`
[C# Eval Function](http://eval-expression.net/) • [SQL Eval Function](http://eval-sql.net/)

**Comment by Bykiev** [comment]
Sure, I'll provide a test project soon.

**Comment by JonathanMagnan** [comment]
Hello @Bykiev ,

A simple reminder that we are here to assist you!

Don't hesitate to provide a sample project, so that we can help you better.

If you need to keep the source private, you can always send it to info@zzzprojects.com

Looking forward to hearing from you, 

Jon

**Comment by Bykiev** [comment]
Hi, sorry for a long delay (deadline is coming), you can find a test sample with Northwind DB in attached archive.

[NMemoryTest.zip](https://github.com/zzzprojects/nmemory/files/5720830/NMemoryTest.zip)

**Comment by JonathanMagnan** [comment]
Hello @Bykiev ,

This error happens when the connection timeout. You can increase the connection as you commented with the `SetConnectionTimeout`.

What happens is there is a rollback and the transaction is aborted.

I will still ask my developer to look at this issue when he will be back from the Christmas break as this is a good scenario to try to optimize the library if we find something.

Best Regards,

Jon

**Comment by Bykiev** [comment]
@JonathanMagnan, hi!

Is any progress on this?

**Comment by JonathanMagnan** [comment]
Hello @Bykiev ,

Sorry for the long wait,

We released today the v2.2.12 which should increase the performance of the insert method.

In some scenarios with thousands of entities, we have seen performance increased by 40%. However, if you still get timeout issue, the only option left on your side will be to increase the timeout.

Let me know if the new version is working fine for your scenario.

Best Regards,

Jon
